File: Graphs/Gold/P1_2016-MooCast/P1_2016-MooCast.cpp
// Graphs - Minimum Spanning Tree (MST) approach
// T: O(n^2), M: O(n), where n is the number of cows

#include <bits/stdc++.h>

using namespace std;

int main() {
	freopen("moocast.in", "r", stdin);
	// First line: n = number of cows
	int n;
	cin >> n;
	// For the next n lines read in coordinates of cows
	vector<pair<int, int>> coords(n);
	for (int idx = 0; idx < n; ++idx)
		cin >> coords[idx].first >> coords[idx].second;

	// Initialize answer variable and distance array (idx of coordinate -> distance)
	vector<int> distance(n, INT_MAX - 1);
	distance[0] = 0;
	int spend = 0;

	// For every cow, find the closest cow and update answer
	for (int cow = 0; cow < n; ++cow) {
		int i = min_element(distance.begin(), distance.end()) - distance.begin();
		spend = max(spend, distance[i]);
		distance[i] = INT_MAX;
		for (int j = 0; j < n; ++j) {
			if (distance[j] == INT_MAX)
				continue;
			int dx = coords[i].first - coords[j].first;
			int dy = coords[i].second - coords[j].second;
			int dist = (dx * dx) + (dy * dy);
			distance[j] = min(distance[j], dist);
		}
	}

	// Write the minimum distance/price to spend to the output file
	freopen("moocast.out", "w", stdout);
	cout << spend << endl;
	return 0;
}

/*
// T: O(n^2)

#include <bits/stdc++.h>
using namespace std;

#define dis(x, y) (p[x].first - p[y].first) * (p[x].first - p[y].first) + (p[x].second - p[y].second) * (p[x].second - p[y].second)

class UF {
private:
    vector<int> parent;

public:
    int count;
    UF(int n) {
        for (int i = 0; i <= n; i++)
            parent.push_back(i);
        count = n; // Initialize count to n
    }

    int find(int x) {
        if (x != parent[x])
            parent[x] = find(parent[x]);
        return parent[x];
    }

    void Union(int x, int y) {
        int rootX = find(x);
        int rootY = find(y);
        if (rootX != rootY) {
            parent[rootX] = rootY;
            count--; // Decrement count only when a merge occurs
        }
    }

    bool connected(int x, int y) {
        int rootX = find(x);
        int rootY = find(y);
        return rootX == rootY;
    }
};

int n;
pair<int, int> p[1001];

bool help(int k) {
    UF u(n);
    for (int i = 1; i <= n; i++) {
        for (int j = i + 1; j <= n; j++) {
            if (dis(i, j) <= k)
                u.Union(i, j);
        }
    }
    return u.count == 1; // Check if all cows are connected
}

int main() {
    freopen("moocast.in", "r", stdin);
    freopen("moocast.out", "w", stdout);
    cin >> n;
    for (int i = 1; i <= n; i++) {
        cin >> p[i].first >> p[i].second;
    }
    int l = 0, r = INT_MAX; // Use a larger range for r
    while (l <= r) {
        int mid = (l + r) / 2;
        if (help(mid))
            r = mid - 1;
        else
            l = mid + 1;
    }
    cout << l << endl;
    return 0;
}
*/

/*
// Graphs - Disjoint Set Union (DSU) approach
// T: O(n^2 * log(n)), M: O(n^2), where n is the number of cows

#include <bits/stdc++.h>

using namespace std;

class DisjointSetUnion {
private:
	vector<int> parent;
	vector<int> size;
public:
	// Initialize nodes to single sized node graphs
	DisjointSetUnion(int n) : size(n, 1) {
		for (int node = 0; node < n; ++node)
			parent.push_back(node);
	}

	// Recursively find and assign parent of current node (path compression)
	int get_parent(int v) {
		return (v == parent[v]) ? v : (parent[v] = get_parent(parent[v]));
	}

	// Attempt to union two components
	bool link(int a, int b) {
		a = get_parent(a), b = get_parent(b);
		if (a == b) // If nodes are already in same component
			return false;

		if (size[a] < size[b]) 
			swap(a, b); // Union by size optimization
		// Merge smaller component (b) into larger component (a)
		parent[b] = a;
		size[a] += size[b];
		return true;
	}
};

int main() {
	freopen("moocast.in", "r", stdin);
	// First line: n = number of cows
	int n;
	cin >> n;
	// For the next n lines read in coordinates of cows
	vector<pair<int, int>> coords(n);
	for (int idx = 0; idx < n; ++idx)
		cin >> coords[idx].first >> coords[idx].second;
	
	// Store edges and calculated distance between cows
	vector<tuple<int, int, int>> edges;
	for (int i = 0; i < n - 1; ++i) {
		// Calculate distance between cow i and all other cows
		for (int j = i + 1; j < n; ++j) {
			int dx = coords[i].first - coords[j].first;
			int dy = coords[i].second - coords[j].second;
			edges.push_back(make_tuple(i, j, (dx * dx) + (dy * dy)));
		}
	}

	// Sort edges by their distance
	sort(edges.begin(), edges.end(), 
	[](tuple<int, int, int> a, tuple<int, int, int> b) {
		return get<2>(a) < get<2>(b);
	});

	// Union closest cows until all cows are connected
	int lastDist = 0;
	DisjointSetUnion dsu(n);
	for (tuple<int, int, int>& e : edges) {
		if (dsu.link(get<0>(e), get<1>(e))) {
			lastDist = get<2>(e);
			if (--n == 1)	// If all cows are connected
				break;
		}
	}

	// The distance of the last unioned edge is the minimum amount to spend
	freopen("moocast.out", "w", stdout);
	cout << lastDist << endl;
	return 0;
}
*/

File: Graphs/Gold/P1_2016-MooCast/P1_2016-MooCast.py
# Graphs - Minimum Spanning Tree (MST) approach
# T: O(n^2), M: O(n), where n is the number of cows

with open("moocast.in", "r") as fin:
	# First line: n = number of cows
	n = int(fin.readline().strip())
	# For the next n lines read in coordinates of cows
	coords = [tuple(map(int, fin.readline().split())) for ln in range(n)]

# Initialize distance hashmap (x, y) -> distance and answer variable
distance = {coord: float("inf") for coord in coords}
distance[coords[0]] = 0
spend = 0

# For every cow, find the closest cow and update answer
for cow in range(n):
	ax, ay = min(distance, key=distance.get)	# Get smallest coordinate
	spend = max(spend, distance[(ax, ay)])
	del distance[(ax, ay)]
	# For every other non-visited/deleted coordinate, update distance
	for bx, by in distance.keys():
		curr_distance = (ax - bx)**2 + (ay - by)**2
		distance[(bx, by)] = min(distance[(bx, by)], curr_distance)

# Write the minimum distance/price to spend to the output file
print(spend, file=open("moocast.out", "w"))

'''
# Graphs - Disjoint Set Union (DSU) approach
# T: O(n^2 * log(n)), M: O(n^2), where n is the number of cows

class DisjointSetUnion:
	# Initialize nodes to single sized node graphs
	def __init__(self, n: int) -> None:
		self.parent = [*range(n)]
		self.size = [1] * n

	# Recursively find and assign parent of current node
	def get_parent(self, v: int) -> int:
		if self.parent[v] == v:
			return v
		# Path compression optimization
		self.parent[v] = self.get_parent(self.parent[v])
		return self.parent[v]
	
	# Attempt to union two components
	def link(self, a: int, b: int) -> bool:
		a, b = self.get_parent(a), self.get_parent(b)
		if a == b:		# If nodes are already in same component
			return False
		
		if self.size[a] < self.size[b]:
			a, b = b, a		# Union by size optimization
		# Merge smaller component (b) into larger component (a)
		self.parent[b] = a
		self.size[a] += self.size[b]
		return True

with open("moocast.in", "r") as fin:
	# First line: n = number of cows
	n = int(fin.readline().strip())
	# For the next n lines read in coordinates of cows
	coords = [tuple(map(int, fin.readline().split())) for ln in range(n)]

# Store edges of indexed coordinates and calculate distanceances between all pairs of cows
edges = [(l_idx, r_idx, (coords[l_idx][0] - coords[r_idx][0])**2 + (coords[l_idx][1] - coords[r_idx][1])**2) for l_idx in range(n - 1) for r_idx in range(l_idx + 1, n)]

# Sort edges by their distanceance
edges.sort(key=lambda x: x[2])

# Union closest cows until all cows are connected
last_distance = 0
dsu = DisjointSetUnion(n)
for l_idx, r_idx, distance in edges:
	if dsu.link(l_idx, r_idx):
		last_distance = distance
		n -= 1
		if n == 1:	# If all cows are connected
			break

# The distanceance of the last unioned edge is the minimum amount to spend
print(last_distance, file=open("moocast.out", "w"))
'''

'''
# Graphs - Binary Search + Breadth-first search (BFS) approach
# T: O(n^2 * log(hi)), M: O(n)
# Where n is the number of cows and hi is the maximum distanceance between any two cows

from collections import deque

with open("moocast.in", "r") as fin:
	# First line: n = number of cows
	n = int(fin.readline().strip())
	# For the next n lines read in coordinates of cows
	coords = [tuple(map(int, fin.readline().split())) for ln in range(n)]

# Breadth-first search (BFS) to find all cows reachable given a distanceance
def all_reachable(spent: int) -> bool:
	q = deque([0])
	seen = set([0])
	while q:
		node = q.popleft()
		nx, ny = coords[node]

		for adj in range(n):
			if adj in seen:
				continue

			ax, ay = coords[adj]
			distanceance = (nx - ax)**2 + (ny - ay)**2
			if distanceance <= spent:
				q.append(adj)
				seen.add(adj)

	return len(seen) == n

# Binary search to find the minimum distanceance/price to spend
min_x, min_y = min(coords)
max_x, max_y = max(coords)
lo, hi = 0, (max_x - min_x)**2 + (max_y - min_y)**2
while lo < hi:
	mid = (lo + hi) // 2
	if (all_reachable(mid)):
		hi = mid
	else:
		lo = mid + 1

# Write binary searched answer to output file
print(lo, file=open("moocast.out", "w"))
'''